### PR TITLE
Make Keras models Pickle-able

### PR DESCRIPTION
**Pull request by farizrahman4u**
#10475

### CODE REVIEW

**Comment by farizrahman4u** [comment]
@fchollet unrelated test seems to be failing?

**Comment by Dref360** [comment]
The CNTK test is a fluke, but the pep8 one is real?

```
/home/travis/build/keras-team/keras/keras/engine/saving.py:80:13: E126 continuation line over-indented for hanging indent
            'class_name': model.__class__.__name__,
            ^
/home/travis/build/keras-team/keras/keras/engine/saving.py:82:9: E121 continuation line under-indented for hanging indent
        }, default=_get_json_type).encode('utf8')
```

File: keras/engine/saving.py [hunk 1]
```diff
@@ -16,13 +17,245 @@
...
+        return obj.__name__
+
+    # if obj is a python 'type'
+    if type(obj).__name__ == type.__name__:
+        return obj.__name__
+
+    raise TypeError('Not JSON Serializable:', obj)
+
+
+def get_model_state(model):
```
Comment (fchollet) [inline keras/engine/saving.py]: This function would exist exclusively for pickling, so I think it should have "pickle" or some such in the name (e.g. `get_model_state_for_pickling`

Additionally, the code seems highly redundant with save_model / load_model. How can we refactor to minimize code duplication?

Comment (farizrahman4u) [reply keras/engine/saving.py]: @fchollet What about - we have a common `get_model_state()`/`load_model_from_state()` which will be called by `pickle_model()`/`unpickle_model()` and `load_model()`/`save_model()`

Comment (fchollet) [reply keras/engine/saving.py]: That sounds good!

**Comment by csbrown** [comment]
I think that it may be possible to use the hdf5 serialization and pickle at the same time.  The trick is to hdf5 serialize the model in memory (which can be done, see `h5py.File` options `driver` and `backing_store`).  Note that the `save_model` method can accept either a string filename or an `h5py.File` object.  Having initialized the appropriate `h5py.File` object, `pickle` is able to serialize this file object.  This can be combined with a `__get_state__` and `__set_state__` method on a base class to perform the appropriate conversions in preparation for pickling.

As a side-benefit of this process, there need not be another materially different serialization method.

Anybody see any pitfalls to this design?

File: keras/engine/saving.py (hunk 1 above)
Comment (farizrahman4u) [reply keras/engine/saving.py]: After further thought, that would make h5py write sub-optimal. We would be creating a fully copy of the model weights in memory and writing it all at once to disk. So probably leave `save_model` as is?

Comment (farizrahman4u) [reply keras/engine/saving.py]: @fchollet ping ping

**Comment by fchollet** [comment]
Hi, I think this is a useful feature and we should merge it. What's the status on this PR? Thanks.

**Comment by farizrahman4u** [comment]
>  What's the status on this PR?

It is functional. 

You had asked me to refactor it such that we have a single serialization method which would be called by `model.save()` and `model.__getstate__()`. But the H5py file is written on the go, and having a common serialization method would make it sub optimal (first a copy of the model will be created in memory and then written to disk). So I have left it as it is.

**Comment by fchollet** [comment]
Seemingly some issues on CI: https://travis-ci.org/keras-team/keras/builds/421708916?utm_source=github_status&utm_medium=notification

File: keras/engine/saving.py [hunk 2]
```diff
@@ -16,13 +17,245 @@
...
+    # if obj is a serializable Keras class instance
+    # e.g. optimizer, layer
+    if hasattr(obj, 'get_config'):
+        return {'class_name': obj.__class__.__name__,
+                'config': obj.get_config()}
+
+    # if obj is any numpy type
+    if type(obj).__module__ == np.__name__:
+        if isinstance(obj, np.ndarray):
+            return {'type': type(obj),
```
Comment (fchollet) [inline keras/engine/saving.py]: We've changed this behavior recently, this will need to be updated

**Comment by fchollet** [comment]
> any chance we can do this without introducing about 95% code duplication for get_json_type etc.?

Is there any possibility that using a shared function with a callback structure would work? In one case the callback would write to a HDF5 file, in the other case it would just build a dict.

**Comment by fchollet** [comment]
Or we create a dict-like HDF5 file class.

**Comment by farizrahman4u** [comment]
@fchollet See #11030 with the dict-like HDF5 file class.

**Comment by fchollet** [comment]
Let's close this PR and move this to #11030.

**Comment by namish800** [comment]
I am having the same problem when I try to run `model.to_json()` my model contains custom layers and I get this error
>TypeError: can't pickle _thread.RLock objects
Is there any workaround this problem?

**Pull request closed without merging**